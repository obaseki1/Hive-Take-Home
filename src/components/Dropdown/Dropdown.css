.main {
  width: 90%;
  max-width: 400px;
  margin: 0 auto;
  padding-top: 100px;
}
.dropdown-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 10px;
  width: 100%;
  box-shadow: 0 1px 4px 1px grey;
  border-radius: 5px;
}
.dropdown {
  width: 100%;
  min-height: 50px;
  max-height: 80px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding-left: 10px;
  padding-right: 10px;
  box-sizing: border-box;
  border-bottom: 0.5px solid grey;
}
.buttons-div {
  width: 100%;
  display: flex;
  justify-content: space-between;
}
.buttons-div span {
  box-shadow: 0 1px 4px 1px grey;
  padding: 6px 4px;
  border-radius: 5px;
  font-weight: 500;
  cursor: pointer;
}
.buttons-div span:hover {
  background-color: rgb(74, 144, 226);
  color: white;
}
.title {
  padding: 4px 0px;
  margin: 10px;
  font-size: 16px;
  font-weight: 500;
  color: grey;
  text-transform: capitalize;
}
.dropdown-values-container {
  width: 100%;
  /*max-height: 500px; 
  overflow-y: scroll;*/
  border-radius: 5px;
}
.dropdown-value-div {
  display: flex;
  justify-content: space-around;
  align-items: center;
  width: 100%;
  text-transform: capitalize;
  text-align: center;
  padding: 4px 0 4px;
  cursor: pointer;
}
.dropdown-value-div:hover {
  background-color: aliceblue;
}
.value-text {
  width: 50%;
  overflow-wrap: break-word;
}
.selected-values-container {
  display: flex;
  flex-wrap: wrap;
  margin: 10px;
  width: 90%;
  max-height: 75px;
  overflow-y: scroll;
}
.selected-value {
  display: flex;
  align-items: center;
  background-color: #e0e0e0;
  padding: 4px 10px;
  margin: 5px;
  border-radius: 5px;
  color: rgba(0, 0, 0, 0.6);
  cursor: pointer;
}
.selected-value-text {
  font-weight: bold;
  font-size: 14px;
  text-transform: capitalize;
}
.selected-value:hover {
  color: black;
}
