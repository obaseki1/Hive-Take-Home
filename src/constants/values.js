const values = [
  'javascript',
  'html',
  'css',
  'react',
  'front end development',
  'redux',
  'mongoDB',
  'express.js',
  'node.js',
  'bootstrap',
  'material ui',
  'responsive design',
  'wordpress',
];

export default values