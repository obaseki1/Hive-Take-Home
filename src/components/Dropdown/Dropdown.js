import React, { useState, useRef } from 'react';
import {
  FaCheckCircle,
  FaChevronUp,
  FaChevronDown,
  FaTimes,
} from 'react-icons/fa';
import {
  List,
  AutoSizer,
  CellMeasurer,
  CellMeasurerCache,
} from 'react-virtualized';
import { useDispatch } from 'react-redux';
import {
  addSkill,
  removeSkill,
  addAllSkills,
  removeAllSkills,
} from '../../actions/skillsActions';
import './Dropdown.css';

function Dropdown({ values, title, selectedValues }) {

  const dispatch = useDispatch();
  const [closed, setClosed] = useState(true);

  const cache = useRef(
    new CellMeasurerCache({
      fixedWidth: true,
      defaultHeight: 40,
    })
  );

  const toggleDropdown = () => {
    values ? setClosed(!closed) : window.alert('No values given');
  };

  const handleValueClick = (selected) => {
    if (selectedValues.indexOf(selected) === -1) {
      dispatch(addSkill(selected));
    } else {
      dispatch(removeSkill(selected));
    }
  };

  const selectAll = () => {
    dispatch(addAllSkills(values));
  };

  const deselectAll = () => {
    dispatch(removeAllSkills());
  };

  const renderRow = ({ index, key, parent, style }) => (
    <CellMeasurer
      key={key}
      cache={cache.current}
      parent={parent}
      columnIndex={0}
      rowIndex={index}
    >
      <div
        className="dropdown-value-div"
        onClick={() => handleValueClick(values[index])}
        style={style}
      >
        {selectedValues.indexOf(values[index]) !== -1 ? (
          <FaCheckCircle color="grey" size="15" />
        ) : (
          <div style={{ width: 15 }}></div>
        )}

        <span className="value-text">{values[index]}</span>
        <div style={{ width: 15 }} />
      </div>
    </CellMeasurer>
  );

  return (
    <div data-testid="dropdown-test" className="main">
      <div className="buttons-div">
        <span onClick={() => selectAll()}>Select All</span>
        <span onClick={() => deselectAll()}>Deselect All</span>
      </div>
      <div className="dropdown-container">
        <div className="dropdown">
          
          {selectedValues.length < 1 ? (
            <span className="title">{title ? title : 'Select Values'}</span>
          ) : (
            <div className="selected-values-container">
              {selectedValues.map((value) => {
                return (
                  <div key={value} className="selected-value">
                    {' '}
                    <span className="selected-value-text">{value}</span>{' '}
                    <FaTimes onClick={() => handleValueClick(value)} />{' '}
                  </div>
                );
              })}
            </div>
          )}

          {closed ? (
            <FaChevronDown
              style={{ cursor: 'pointer' }}
              size="16"
              onClick={() => toggleDropdown()}
            />
          ) : (
            <FaChevronUp
              style={{ cursor: 'pointer' }}
              size="16"
              onClick={() => toggleDropdown()}
            />
          )}
        </div>

        {!closed && (
          <div
            className="dropdown-values-container"
            onBlur={() => toggleDropdown()}
          >
            <div style={{ width: '100%', height: 300 }}>
              <AutoSizer>
                {({ width, height }) => (
                  <List
                    width={width}
                    height={height}
                    rowRenderer={renderRow}
                    rowCount={values.length}
                    rowHeight={50}
                    deferredMeasurementCache={cache.current}
                  />
                )}
              </AutoSizer>
            </div>
          </div>
        )}
      </div>
    </div>
  );
}

export default Dropdown;
