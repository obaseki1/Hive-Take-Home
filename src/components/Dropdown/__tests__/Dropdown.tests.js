import React from 'react';
import Dropdown from '../Dropdown';
import { render, screen } from '../../../test-utils';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import values from '../../../constants/values';

const skills = ['test', 'test1', 'test2'];

test('renders without crashing', () => {
  const div = document.createElement('div');
  render(<Dropdown values={values} selectedValues={skills} title="test"/>, div);
});

test('matches snapshot', () => {
  const div = document.createElement('div');
    const { getByTestId } = render(<Dropdown values={values} selectedValues={skills} title="Testing"/>, div);
    expect(getByTestId('dropdown-test')).toMatchSnapshot()
});
