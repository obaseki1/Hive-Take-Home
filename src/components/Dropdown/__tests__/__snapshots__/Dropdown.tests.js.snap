// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="main"
  data-testid="dropdown-test"
>
  <div
    class="buttons-div"
  >
    <span>
      Select All
    </span>
    <span>
      Deselect All
    </span>
  </div>
  <div
    class="dropdown-container"
  >
    <div
      class="dropdown"
    >
      <div
        class="selected-values-container"
      >
        <div
          class="selected-value"
        >
           
          <span
            class="selected-value-text"
          >
            test
          </span>
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 352 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
           
        </div>
        <div
          class="selected-value"
        >
           
          <span
            class="selected-value-text"
          >
            test1
          </span>
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 352 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
           
        </div>
        <div
          class="selected-value"
        >
           
          <span
            class="selected-value-text"
          >
            test2
          </span>
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 352 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
           
        </div>
      </div>
      <svg
        fill="currentColor"
        height="16"
        stroke="currentColor"
        stroke-width="0"
        style="cursor: pointer;"
        viewBox="0 0 448 512"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </div>
  </div>
</div>
`;
